### PR TITLE
decorator to deprecate positional arguments

### PR DESCRIPTION
**Pull request by mathause**
- [x] Supersedes  #6403, see also #5531
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

Adds a helper function to deprecate positional arguments. IMHO this offers a good trade-off between magic and complexity. (As mentioned this was adapted from scikit-learn).

edit: I suggest to actually deprecate positional arguments in another PR.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
Are the functions you are considering using this functions that never had keyword arguments before? When I wrote a similar decorator before i had an explicit list of arguments that were allowed to be converted.

**Comment by headtr1ck** [comment]
Do you think that you can get it to work with static typing?

**Comment by mathause** [comment]
> Are the functions you are considering using this functions that never had keyword arguments before? When I wrote a similar decorator before i had an explicit list of arguments that were allowed to be converted.

No, I want to consider arguments that have default values.

```python
def mean(dim, skipna=None, keep_attrs=None):
    pass

@_deprecate_positional_args("v0.1.0")
def mean(dim, *, skipna=None, keep_attrs=None):
    pass
```

Or do I missunderstand your question?

> Do you think that you can get it to work with static typing?

I think `functools.wraps` also attaches the annotations?

**Comment by mathause** [comment]
I just realized that there is a bug in the decorator for "keyword-only arguments without a default". I wonder if I should fix this or just disallow this pattern, i.e. the following does currently not work properly:

```python
@_deprecate_positional_args("v0.1.0")
def func(a, *, b):
    pass
```

but I am not sure how helpful pattern is.

**Comment by mathause** [comment]
I also just realized, that we made many arguments keyword-only without deprecation (e.g. `xr.Dataset.mean`) - but that should not stop us from using this - there are still a lot of instances where it could be helpful.

**Comment by mathause** [comment]
Any objections to merge this as is?

**Comment by dcherian** [comment]
:+1: put it on the meeting agenda for today

File: xarray/util/deprecation_helpers.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                    # if removing this constraint -> need to add these to kwargs as well
+                    raise TypeError("Keyword-only param without default disallowed.")
+            elif param.kind == POSITIONAL_ONLY:
+                raise TypeError("Cannot handle positional-only params")
+                # because all args are coverted to kwargs below
+
+        @wraps(f)
+        def inner(*args, **kwargs):
+            print(f"{args=}")
+            print(f"{pos_or_kw_args=}")
```
Comment (shoyer) [inline xarray/util/deprecation_helpers.py:63]: are these print statements intentional?

File: xarray/util/deprecation_helpers.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import inspect
```
Comment (shoyer) [inline xarray/util/deprecation_helpers.py:1]: I would suggest keeping the original copyright notice at the top of the this file.

File: xarray/util/deprecation_helpers.py (hunk 1 above)
Comment (dcherian) [reply xarray/util/deprecation_helpers.py:96]: @mathause I think you missed these print statements?

Comment (mathause) [reply xarray/util/deprecation_helpers.py:96]: Thanks! Will fix them in a follow up!

**Pull request merged**